File: django-backend/api/views/post.py
from typing import List

from ninja import Router, Form, File, UploadedFile
from ninja.pagination import paginate
from django.db.models import Q
from pydantic import UUID4

from api.models import User, Post, PostFile, Notification, PostLike, PostComment
from api.schemas import PostIn, PostOutMinimal, PostOut, CommentOutMinimal, CommentOut
from api.helpers import save_file, create_notification


router = Router(tags=["post"])


@router.post("", response={201: PostOutMinimal, 422: str})
def create_post(request, data: Form[PostIn], files: File[list[UploadedFile]] = []):
    if not data.content and not files:
        return 422, "Data not provided"

    post = Post.objects.create(author=request.auth, content=data.content)
    post_files = PostFile.objects.bulk_create([PostFile(post=post) for _ in files])

    for pf, file in zip(post_files, files):
        save_file(pf.file, file, ["image", "video"])

    for friend_id in request.auth.friends():
        create_notification(
            friend_id,
            Notification.Types.FRIEND_POSTED,
            {
                "postId": str(post.id),
                "userId": str(request.auth.id),
                "firstName": request.auth.first_name,
                "lastName": request.auth.last_name,
                "profileImage": (
                    request.auth.profile_image.url
                    if request.auth.profile_image
                    else None
                ),
            },
        )

    return 201, post


@router.get("/feed", response={200: List[PostOut]})
@paginate
def get_feed(request):
    return Post.include_extra(
        Post.objects.filter(
            Q(author__id__in=request.auth.friends()) | Q(author=request.auth)
        ),
        request.auth,
    ).order_by("-posted_at")


@router.post("/like/{id}", response={404: str, 201: str, 200: str})
def like_post(request, id: UUID4):
    post = Post.objects.filter(id=id).first()
    if not post:
        return 404, "Post not found"

    post_liked = PostLike.objects.filter(post=post, liked_by=request.auth)
    if post_liked:
        post_liked.delete()
        return 200, "Post unliked"
    else:
        PostLike.objects.create(post=post, liked_by=request.auth)
        create_notification(
            post.author.id,
            Notification.Types.POST_LIKED,
            {
                "postId": str(post.id),
                "userId": str(request.auth.id),
                "firstName": request.auth.first_name,
                "lastName": request.auth.last_name,
                "profileImage": (
                    request.auth.profile_image.url
                    if request.auth.profile_image
                    else None
                ),
            },
        )
        return 201, "Post liked"


@router.post("/comment/{id}", response={201: CommentOutMinimal, 404: str, 422: str})
def comment_post(
    request, id: UUID4, data: Form[PostIn], file: File[UploadedFile] = None
):
    if not data.content and not file:
        return 422, "Data not provided"

    post = Post.objects.filter(id=id).first()

    if not post:
        return 404, "Post not found"

    comment = PostComment.objects.create(
        post=post, author=request.auth, content=data.content
    )
    save_file(comment.file, file, ["image", "video"])

    create_notification(
        post.author.id,
        Notification.Types.POST_COMMENTED,
        {
            "postId": str(post.id),
            "userId": str(request.auth.id),
            "firstName": request.auth.first_name,
            "lastName": request.auth.last_name,
            "profileImage": (
                request.auth.profile_image.url if request.auth.profile_image else None
            ),
            "content": data.content or "An attachment",
        },
    )
    return 201, comment


@router.delete("/comment/{id}", response={200: str, 404: str})
def delete_comment(request, id: UUID4):
    comment = PostComment.objects.filter(id=id, author=request.auth).first()
    if not comment:
        return 404, "Comment not found"

    comment.delete()
    return 200, "Comment deleted"


@router.get("/{id}/comments", response={404: str, 200: List[CommentOut]})
@paginate
def get_comments(request, id: UUID4):
    post = Post.objects.filter(id=id).first()
    if not post:
        return 404, "Post not found"

    return PostComment.objects.filter(post=post).order_by("-commented_at").select_related("author")


@router.get("/ofUser/{id}", response={404: str, 200: List[PostOut]})
@paginate
def get_user_posts(request, id: UUID4):
    user = User.objects.filter(id=id).first()
    if not user:
        return 404, "User not found"

    return Post.include_extra(Post.objects.filter(author=user).order_by("-posted_at"), request.auth)


@router.get("/{id}", response={404: str, 200: PostOut})
def get_post(request, id: UUID4):
    return Post.include_extra(Post.objects.filter(id=id), request.auth).first() or (
        404,
        "Post not found",
    )


@router.delete("/{id}", response={200: str, 404: str})
def delete_post(request, id: UUID4):
    post = Post.objects.filter(id=id, author=request.auth).first()
    if not post:
        return 404, "Post not found"

    post.delete()
    return 200, "Post deleted"

File: express-backend/src/views/post.js
import fs from "fs/promises";
import { NotificationType } from "@prisma/client";
import express from "express";
import { matchedData } from "express-validator";
import cuid from "cuid";

import * as validator from "../validators/post.js";
import { isValidUser } from "../validators/user.js";
import {
    getPublicFileDirectory,
    getPublicFileURL,
    paginate,
} from "../helpers.js";
import prisma, {
    selectJoinedPostData,
    processFetchedJoinedPostData,
    friendsOf,
} from "../db.js";
import { createNotification } from "./notification.js";

const router = express.Router();

router.post("/", validator.postCreate, async (req, res) => {
    const post = await prisma.post.create({
        data: {
            content: matchedData(req).content,
            authorId: req.user.id,
        },
    });

    await prisma.postFile.createMany({
        data: (req.files || []).map((file) => {
            file.id = cuid();
            const fileName = `${file.id}.${file.mimetype.split("/")[1]}`;
            fs.writeFile(getPublicFileDirectory(fileName), file.buffer);
            return {
                id: file.id,
                postId: post.id,
                file: fileName,
            };
        }),
    });

    const friends = await friendsOf(req.user);
    const userPublicImage = getPublicFileURL(req.user.profileImage);
    friends.forEach((friendId) => {
        createNotification(friendId, NotificationType.FRIEND_POSTED, {
            postId: post.id,
            userId: req.user.id,
            firstName: req.user.firstName,
            lastName: req.user.lastName,
            profileImage: userPublicImage,
        });
    });
    return res.status(201).json({ id: post.id });
});

router.delete("/:id", validator.postExists, async (req, res) => {
    if (req.post.authorId !== req.user.id) return res.sendStatus(403);

    const files = await prisma.postFile.findMany({
        where: { postId: req.params.id },
    });

    files.forEach((file) => fs.unlink(getPublicFileDirectory(file.file)));
    await prisma.post.delete({ where: { id: req.params.id } });
    return res.sendStatus(200);
});

router.post("/like/:id", async (req, res) => {
    const post = await prisma.post.findUnique({
        where: { id: req.params.id },
        include: {
            likes: {
                where: {
                    likedById: req.user.id,
                },
            },
        },
    });

    if (!post) return res.sendStatus(404);

    const likeData = { postId: req.params.id, likedById: req.user.id };

    if (post.likes.length > 0) {
        await prisma.postLike.delete({ where: { likedById_postId: likeData } });
        return res.sendStatus(200);
    } else {
        await prisma.postLike.create({ data: likeData });
        await createNotification(post.authorId, NotificationType.POST_LIKED, {
            postId: req.params.id,
            userId: req.user.id,
            firstName: req.user.firstName,
            lastName: req.user.lastName,
            profileImage: getPublicFileURL(req.user.profileImage),
        });
        return res.sendStatus(201);
    }
});

router.post("/comment/:id", validator.commentCreate, async (req, res) => {
    const file = req.file;
    const content = matchedData(req).content;

    if (!content && !file) return res.sendStatus(400);

    const commentData = {
        content,
        postId: req.params.id,
        authorId: req.user.id,
    };

    if (file) {
        const fileName = `${cuid()}.${file.mimetype.split("/")[1]}`;
        fs.writeFile(getPublicFileDirectory(fileName), file.buffer);
        commentData.file = fileName;
    }

    const comment = await prisma.postComment.create({ data: commentData });
    await createNotification(
        req.post.authorId,
        NotificationType.POST_COMMENTED,
        {
            postId: req.params.id,
            userId: req.user.id,
            firstName: req.user.firstName,
            lastName: req.user.lastName,
            profileImage: getPublicFileURL(req.user.profileImage),
            content: content || "An attachment",
        }
    );
    return res.status(201).json({ id: comment.id });
});

router.delete("/comment/:id", validator.commentExists, async (req, res) => {
    if (req.comment.authorId !== req.user.id) return res.sendStatus(403);

    if (req.comment.file) fs.unlink(getPublicFileDirectory(req.comment.file));
    await prisma.postComment.delete({ where: { id: req.params.id } });
    return res.sendStatus(200);
});

router.get("/:id/comments", validator.postExists, async (req, res) => {
    const comments = await prisma.postComment.findMany({
        where: {
            postId: req.params.id,
        },
        include: {
            author: true,
        },
        orderBy: { commentedAt: "desc" },
        ...paginate(req),
    });

    for (let comment of comments) {
        comment.file = getPublicFileURL(comment.file);
        comment.author.profileImage = getPublicFileURL(
            comment.author.profileImage
        );
    }

    return res.json({ items: comments });
});

router.get("/feed", async (req, res) => {
    return res.json({
        items: (
            (await prisma.post.findMany({
                where: {
                    authorId: {
                        in: [...(await friendsOf(req.user)), req.user.id],
                    },
                },
                orderBy: {
                    postedAt: "desc",
                },
                ...selectJoinedPostData(req),
                ...paginate(req),
            })) || []
        ).map((p) => processFetchedJoinedPostData(p)),
    });
});

router.get("/ofUser/:id", isValidUser(), async (req, res) => {
    return res.json({
        items: (
            (await prisma.post.findMany({
                where: { authorId: req.params.id },
                orderBy: { postedAt: "desc" },
                ...selectJoinedPostData(req),
                ...paginate(req),
            })) || []
        ).map((p) => processFetchedJoinedPostData(p)),
    });
});

router.get("/:id", validator.postExists, async (req, res) => {
    let post = await prisma.post.findUnique({
        where: { id: req.params.id },
        ...selectJoinedPostData(req),
    });

    return res.json(processFetchedJoinedPostData(post));
});

export default router;
